### DISCUSSION TITLE
Vim - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#99] [Announcements]
# Vim - CS自学指南

CS自学指南

https://csdiy.wiki/%E5%BF%85%E5%AD%A6%E5%B7%A5%E5%85%B7/Vim/

**Comment by xiaofan2333** [comment]
建议从missing semeter听讲开始学，不然很折磨

**Reply by Flora-2017** [reply]
You are right

**Reply by MTDickens** [reply]
现在 csdiy 里有更好的课了，可以在 semester 的附近找到。

**Reply by MandyZhang17** [reply]
求问semester附近指的是？

**Reply by liyuelian** [reply]
应该说的是 UCB: Sysadmin DeCal

**Reply by liyuelian** [reply]
另外推荐看：The linux Command Line
中文版：https://www.kancloud.cn/thinkphp/linux-command-line

**Reply by MandyZhang17** [reply]
谢谢楼上的佬！

**Comment by hyu17** [comment]
在VSCode上安装了Vim插件之后效率提高了不少

**Comment by smallst** [comment]
只有vim没有emacs不开心， 有空要pr一下了

**Comment by imageslr** [comment]
分享一下我的自学笔记：https://imageslr.com/2021/vim.html

**Reply by NYfLAmE** [reply]
写得好，赞一个

**Reply by imageslr** [reply]
做了个视频：https://www.bilibili.com/video/BV1MX4y1b7nM

**Reply by Dongmayyys** [reply]
写的真仔细

**Reply by Chimsvov** [reply]
网站做的真好

**Comment by eli-yip** [comment]
补充一下关于mac上的键位映射：
个人建议只将CapsLock映射到Esc，而不是按住为Ctrl，点按位Esc，实测本文中提供的脚本会导致快速按CapsLock和其他键的时候另一个按键有一定几率不被vim捕获，比如，我想从insert mode切换到normal mode，再新建一行并进入insert mode，按键顺序是CapsLock->O，由于敲击速度过快，很可能只退出了insert mode，而没有新建一行。
虽然几率不大，但是这个操作很常用，不确定性会令人很恼火。
供参考。

**Reply by GabeGu** [reply]
个人觉得离得较远的键位，可以适时改变下长时间保持的手部姿势，缓解缓解走向腕管综合征的进度，哈哈~~

**Reply by AugustineFu** [reply]
但是这样的话ctrl就会很难按，我一直苦恼于怎么设置<esc>和<ctrl>

**Reply by smallst** [reply]
@AugustineFu 试试 caps 是control，jj是 esc，或者kj是esc

**Comment by keepkeen** [comment]
上了一学期的linux课，一直以为叫vi面板，原来这就是vim编辑器啊

**Comment by sherlcok314159** [comment]
分享一下自己的经验：将Esc -> jj这样比较舒服:-)

**Reply by MauveNightshade** [reply]
我是把jj jk kj kk全部映射为Esc

**Comment by GabeGu** [comment]
不用管陡峭不陡峭，知道它最终会给自己带来回报并且用就是了~(虽然一开始会变慢，但内心是笃定的，我会越来越快！) 最终发现自己学会的不仅仅是用vim方式编辑，让自己的思路化作一次次流畅丝滑地敲击，更重要的是DRY(Do not Repeat Yourself)的思维，在生活的方方面面你都会带着这个认知去实践。

本人的学习路径是：
vimtutor -> vim方式编辑 -> 形成肌肉记忆(以前学盲打不就是这个过程嘛~~不是陡峭其实是长~) ->  《Practical Vim》学习技巧 -> vim方式编辑 -> 离不开vim方式编辑(即使没有vim模式，就用基本的快捷键方式移动操作)。

这本指南看到这里，看着左侧展开的目录栏，作为毕业十年，非科班出身但循着对计算机科学世界的好奇，在工作中及之余摸索着一块一块地拾起这座大厦的拼图的自己，此时此刻内心起了波澜，很多感叹。
想象到将有更多的富有创造力和激情的后辈们有如此好的起点，倍感欣慰。同时向作者致敬！

**Comment by unuing** [comment]
> 但更佳的做法是同时将 CapsLock 映射为 Ctrl 和 Esc，点按为 Esc，按住为 Ctrl

其实只要用`ctrl`+`[`进normal mode就不用这么麻烦了

**Comment by LunarianCubed** [comment]
推荐 `Learn Vim (the Smart Way)`
有中文翻译

**Comment by Rogerskelamen** [comment]
不害臊地推荐一下[自己的Neovim配置](https://github.com/Rogerskelamen/nvim.lua)😜

**Comment by We-unite** [comment]
键盘映射除了可以用AutoHotkey设置快捷键之外，如果只是需要固定键直接换位置（比如直接把Esc和CapsLock简单换个位置），可以用ScanCode Map的方法在注册表里改。这种方法的优点是：

+ 基于注册表会使得**只要开机就运行**（比如开机之后输入密码之前，CapsLock也出现在Esc的位置上），而不是AutoHotkey脚本需要设置开机启动还要一直运行，~~多少算是能省下一点算力吧~~
+ 由于基于注册表，偏底层，使得该方法**对整个机子都有效**（包括虚拟机，也就是说主机设置之后虚拟机不需要另外设置就有这个键盘映射），而AutoHotkey只对宿主机有效，虚拟机需要另外设置

我的键盘映射非常简单：[放在这里了](https://blog.csdn.net/QJGZ20020910/article/details/129029893?spm=1001.2014.3001.5502)

**Comment by Schroedingersdoraemon** [comment]
分享一下我个人的方案：交换 ESC 和 CapsLock，空格右边的 Alt 映射为 Ctrl
另外推荐一下 kmonad，可在 win macos linux 下使用，其中 linux 支持 tty x11 wayland :)

**Reply by Schroedingersdoraemon** [reply]
[kmonad](https://github.com/kmonad/kmonad)

**Reply by SOV710** [reply]
我用的是keyd，也很快，但是和tmux兼容需要修一修，也支持wayland :)

**Comment by RyanGao1995** [comment]
https://vim-adventures.com/
（非广告）这是一个让你熟悉VIM的游戏，25美元。这能让你很快熟悉一些VIM的操作。
（你需要半年还是90天完成，不然就要再买一次😂）

**Comment by CainLin219** [comment]
真的是好细致无私的指南，狂赞

**Comment by Mc-GrowlR** [comment]
vim在b站有个TheCW的up主讲的不错

**Reply by imageslr** [reply]
我也看过。老哥总爱摸头，讲得确实不错。不过太酷炫了，要配不少插件。入门还是先把指令学好吧。

**Reply by Mc-GrowlR** [reply]
刚才才接到我的邮件提示，有点无语，快一年了。
总而言之，现在可以用nvim啦，有不少的预方案，我现在用的就是LazyVim。
个人经验是了解了一些基本概念之后就可以上手了。

**Comment by TephrocactusMYC** [comment]
推荐一下最新的Neovim，需要注意的是网上很多资料都是基于vim7.x以前的版本，而8.x而最新的9.x，以及Neovim有很多强大的新特性和插件，如果是经常在远程环境或者某些嵌入式环境下编程的朋友可以考虑，不过那些高度定制化的vim/neovim往往只适用于某一种语言，比如C++，如果你的环境经常编辑C++、python、JS、HTML、Rust等多种语言，那可能达不到预期效果，个人的一个小教训。如果电脑配置足够，可以每种编程语言专门配置一个环境（虚拟机），然后再高度定制自己的vim用来充当这种语言的IDE。由于本人能力有限，因此说的不一定正确，欢迎批评指正。

**Reply by TephrocactusMYC** [reply]
怎么说呢，凡是那种拿Vim当IDE用的，感觉用途都变窄了，如果你不是专注于某一种语言的，我不太建议把Vim配成IDE（毕竟你不会用Pycharm写JAVA吧），可以尝试把Neovim配置成你喜欢的IDE，然后把Vim、VSCODE、EMACS的vim模式都当作vim来用，是不错的选择。

**Comment by happytech-web** [comment]
我是准大一学生，大家建议先学完必学工具还是先学数学基础，现在连这个软件怎么用啥时候用都不知道

**Reply by smallst** [reply]
先看 https://github.com/PKUFlyingPig/cs-self-learning/blob/master/docs/CS%E5%AD%A6%E4%B9%A0%E8%A7%84%E5%88%92.md
和 https://github.com/PKUFlyingPig/cs-self-learning/blob/master/docs/%E4%BD%BF%E7%94%A8%E6%8C%87%E5%8D%97.md

**Comment by Winlchance** [comment]
这玩意还是用的时候再花两三天学，要不然学的没目的很难受

**Reply by happytech-web** [reply]
感谢解惑😀😀

**Comment by blameitonme1** [comment]
编程方面直接使用wsl可行吗，既可以熟悉linux的操作，而且我感觉linux包管理什么的都很方便。

**Reply by smallst** [reply]
可以。但更建议先用完整的linux学linux然后用wsl来编程。 因为wsl是不完整的（魔改的）。

**Comment by Joye-Yu** [comment]
shell里输入vimtutor，自带vim中文教程帮助你快速上手基础操作

**Comment by jeefies** [comment]
分享一个个人觉得写的很好的教程：https://github.com/wsdjeg/Learn-Vim_zh_cn
作者迁移其到了 [gitlab](https://gitlab.com/wsdjeg/Learn-Vim_zh_cn)

**Comment by sansanyuchen** [comment]
谢谢大哥

**Comment by cgwer** [comment]
vimtutor 真的很棒

**Comment by Smart-SangGe** [comment]
推荐使用neovim+lazyvim/lunarvim。没必要为了学vim而用vim

**Comment by goahead656** [comment]
github搜索dotfiles查找配置文件，有很多人开源了他们的配置文件

**Comment by yejunyu** [comment]
1

**Comment by qwang1016** [comment]
windows 用sharpkeys键位映射也很方便

**Comment by Lengyu08** [comment]
https://shuttle.gt-all.com/aff.php?aff=7543
嘎嘎快，速度可以达到链路的上限

**Comment by AI-sudo6** [comment]
求助：上一个工具总是订阅失败，尝试了常见问题里面的方法也不太行

**Comment by F1Justin** [comment]
推荐macOS下修改Caps的脚本使用这个https://ke-complex-modifications.pqrs.org/#e0da_caps_lock，这样可以通过 Shift + Caps 实现大写锁定而不是把这个功能直接丢了

**Reply by F1Justin** [reply]
重新发一下链接（）
https://ke-complex-modifications.pqrs.org/#e0da_caps_lock

**Comment by littledgg** [comment]
如果是ssh连接服务器时，需要在终端修改各种配置文件，那么相比于vi/vim，nano，个人更建议使用micro，和windows完全一样的快捷键，没有学习成本。

**Comment by SOV710** [comment]
作为一个已经把 neovim 作为主力编辑器的人来说，我倒觉得 vim 并不是一个所谓的"必学工具"。当然作为 linux 的一部分，基本的 vi 知识，比如 vim 有几种 mode，各自有什么功能，如何切换 mode，如何退出 vim ，hjlk 移动，y 复制 p 粘贴 / 搜索这些是要学的 (作为 ssh 调试的一部分不得不学) 。而且说实话 vim 的操作其实不难，键盘精灵罢了。但是 vim 用到后面要整天整天鼓捣 vimscript 或 lua 插件，或者天天在 vscode 里面调 json 解决映射冲突，这些才是非常折腾并且折磨人耐心的事情 ( emacs 同理) 。

我觉得，如果你是那种 IDE 用得很六，或者你是 VSCode 高手，只是因为某天看了这篇指南，所以心血来潮想全套复制这套"必学工具"的人，那我只能说你先别急。我推荐你先装个Vscode Vim / Neovim插件，或者装个 IdeaVim 试试手感，用熟了之后自己对比一下再下决定。

如果你适应了 vim 的全键盘操作并且认为这很好，但是你觉得这除了个界面以外啥都没有，没有语法高亮没有自动补全没有debug模式背景又丑，压根敲不了代码，那很好你已经入门了，emacs (lisp) / vim (vimscript) / neovim (lua) 三选一 (括号里指插件所使用的语言)。关于三种编辑器的对比可自行查询。接下来就是自定义编辑器不得不品的配插件环节，因为我是neovim主力用户所以我只给neovim的相关建议。

关于neovim的建议: 如果你不想把时间花在庞杂的各种插件配置上，那么我建议 lazyvim/lunarvim/astrovim/nvchad 四种neovim"发行版"四选一就行了，基本能满足开箱即用的体验。如果你想从零开始自定义自己的一套编辑器，那么github上的**kickstart.nvim**和你是一样想的，并且为你提供了一套模板和油管上的配套视频供你学习 https://youtu.be/m8C0Cq9Uv9o?si=rZX4BFkvvRFwcr55 ，b站上也有搬运，除此之外油管有一大坨neovim教程供你参考。2025年我个人推荐这个更新的视频 https://youtu.be/KYDG3AHgYEs?si=_dLb0ao20fw8hckX ，这个教程帮了我很大的忙。

什么？你说vscode + neovim？~~路边一条（bushi~~

ps: 从"试试手"到"用熟了"也不是轻松的过程，如果你觉得现在的 ide/editor 就非常适合自己，那我觉得还是别换好了。但如果你和我一样认为命令行编程是一种美学，或者追求极致的定制化和可扩展性，那我觉得 vim 确实是一个"必学工具"。

**Reply by Bluky-BlueSky** [reply]
vscode + neovim ~~怎么你了(╯‵□′)╯︵┻━┻（bushi~~

**Comment by Future2486** [comment]
2025年2月4日，通过douyin发现这个宝藏网站，如果其他人有交流群或者社群拿拉我一下吗？第一次学CS希望大家多多关照。3476724212@qq.com

**Reply by Kipchoge963** [reply]
COOL~

**Reply by lzwwwwa** [reply]
有群的话拉一下我，我也是第一次学，感觉好多东西需要补，希望可以一起学习讨论一下

**Comment by Future2486** [comment]
您好，很抱歉现在才看到这个信息，我最近有加了一个QQ的cs自学交流群，你可以加我QQ:3476724212或者vx：future2486



---原始邮件---
发件人: "Ziwei ***@***.***&gt;
发送时间: 2025年2月9日(周日) 下午4:43
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [PKUFlyingPig/cs-self-learning] Vim - CS自学指南 (Discussion #99)




 
有群的话拉一下我，我也是第一次学，感觉好多东西需要补，希望可以一起学习讨论一下

**Comment by zzzfreeuser** [comment]
求markdown的教程

**Reply by tianyindique** [reply]
markdown的教程网上一搜一大把，语法并不难，常用的就是那几个，你随便看看，用的时候再查就是了

**Comment by Bluky-BlueSky** [comment]
我用的是windows环境的vscode+neovim，所以有在win系统下只在vscode中将capslock映射为esc和ctrl，其他软件和系统场景下不变的需求，于是借助gpt5基于[这个脚本](https://gist.github.com/108anup/ecd891f21dc88ef107ff6c8ab0969700)修改了使用ahk v2.0的脚本，如果需要开机自启就把下面的脚本放到这个目录中`C:\Users\用户名\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Startup\caps_to_esc_or_ctrl_for_vscode.ahk` ，如果需要在其他场景下或者全局场景下进行映射可参照脚本注释进行对应修改。

```ahk
#Requires AutoHotkey v2.0
#SingleInstance Force

global lastDown := 0, abortEsc := false

; ---------- 只在 VS Code 生效 ----------
#HotIf WinActive("ahk_exe Code.exe")

*CapsLock:: {
    global lastDown, abortEsc
    Send "{Ctrl down}"
    lastDown := A_TickCount
    abortEsc := false

    ; 使用输入钩子监听 CapsLock 按下期间的其它按键
    ih := InputHook("L1 V", "{All}")
    ih.KeyOpt("{All}", "E")  ; 监听所有键
    ih.OnKey := (*) => abortEsc := true
    ih.Start()
}

*CapsLock Up:: {
    global lastDown, abortEsc
    Send "{Ctrl up}"
    if (!abortEsc && (A_TickCount - lastDown <= 500))
        Send "{Esc}"
}

#HotIf  ; ------- 热键条件结束 -------


; ---------- 在 VS Code 之外 ----------
CapsLock::CapsLock

```

**Comment by DetermiNemo** [comment]
加减等号还有中括号大括号都在键盘右上角，各位都是直接用小拇指按的吗，有没有什么映射方法？

**Comment by zzzfreeuser** [comment]
overleaf支持vim吗

**Comment by lil-Journey** [comment]
自建日本节点服务器 月租5元不限流量 本人也在用 失效时可联系人工修复 qq:1165116287

**Comment by L20wang** [comment]
现在vscode也可以ssh了，不是必需项

**Comment by Vulpes1agopus** [comment]
感觉不管是vscode还是nano都更好用一点

**Comment by karenepitaya** [comment]
Maybe useful: https://gist.github.com/karenepitaya/46aeb717b3202c8d371d50b34eefb568
> I made an ah2 implementation and it works fine in my machine, it's maybe useful for you
